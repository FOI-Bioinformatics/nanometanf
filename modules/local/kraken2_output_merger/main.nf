process KRAKEN2_OUTPUT_MERGER {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/python:3.11' :
        'quay.io/biocontainers/python:3.11' }"

    input:
    tuple val(meta), path(kraken2_outputs)
    path  batch_metadata_files

    output:
    tuple val(meta), path("*.cumulative.kraken2.output.txt"), emit: cumulative_output
    tuple val(meta), path("merge_stats.json"), emit: stats
    path  "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    #!/usr/bin/env python3

    import json
    import sys
    from pathlib import Path

    # Load all batch metadata files
    metadata = []
    metadata_files = "${batch_metadata_files}".split()

    for f in metadata_files:
        with open(f) as fh:
            metadata.append(json.load(fh))

    # Sort by batch_id to ensure correct order
    metadata.sort(key=lambda x: x['batch_id'])

    print(f"Merging {len(metadata)} batch outputs for sample ${meta.id}", file=sys.stderr)

    # Concatenate outputs in batch order
    total_reads = 0
    with open('${prefix}.cumulative.kraken2.output.txt', 'w') as out:
        for m in metadata:
            # Find the corresponding output file
            output_basename = Path(m['kraken2_output']).name

            # Search for the file in the current directory
            output_file = None
            for f in Path('.').glob('*.kraken2.output.txt'):
                if f.name == output_basename or \
                   f.name.startswith(f"${meta.id}_batch{m['batch_id']}"):
                    output_file = f
                    break

            if output_file is None:
                print(f"WARNING: Could not find output for batch {m['batch_id']}", file=sys.stderr)
                continue

            print(f"  Batch {m['batch_id']}: {m['classification_statistics']['total_sequences']} reads", file=sys.stderr)

            with open(output_file) as f:
                lines = f.readlines()
                out.writelines(lines)
                total_reads += len(lines)

    # Generate merge statistics
    merge_stats = {
        'sample_id': '${meta.id}',
        'total_batches': len(metadata),
        'total_reads': total_reads,
        'batches': [m['batch_id'] for m in metadata],
        'cumulative_output': '${prefix}.cumulative.kraken2.output.txt'
    }

    with open('merge_stats.json', 'w') as stats:
        json.dump(merge_stats, stats, indent=2)

    print(f"", file=sys.stderr)
    print(f"Merge complete: {total_reads} total reads from {len(metadata)} batches", file=sys.stderr)

    # Generate versions.yml
    with open('versions.yml', 'w') as v:
        v.write('"${task.process}":\n')
        v.write(f'    python: {sys.version.split()[0]}\n')
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.cumulative.kraken2.output.txt
    echo '{"sample_id": "${meta.id}", "total_batches": 0, "total_reads": 0}' > merge_stats.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | sed 's/Python //g')
    END_VERSIONS
    """
}
