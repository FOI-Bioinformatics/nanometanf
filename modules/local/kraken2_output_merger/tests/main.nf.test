nextflow_process {
    name "Test Process KRAKEN2_OUTPUT_MERGER"
    script "../main.nf"
    process "KRAKEN2_OUTPUT_MERGER"
    tag "modules"
    tag "modules_local"
    tag "kraken2_output_merger"
    tag "kraken2"
    tag "incremental"

    test("merge 2 batches - single sample - stub") {
        options "-stub"

        when {
            process {
                """
                // Create mock batch outputs
                def batch0_output = file("test_batch0.kraken2.output.txt")
                batch0_output.text = "C\\tread1\\t12345\\t100\\t12345:1\\nU\\tread2\\t0\\t100\\t0:100\\n"

                def batch1_output = file("test_batch1.kraken2.output.txt")
                batch1_output.text = "C\\tread3\\t67890\\t100\\t67890:1\\nC\\tread4\\t12345\\t100\\t12345:1\\n"

                input[0] = [
                    [ id:'sample1' ],
                    [ batch0_output, batch1_output ]
                ]

                // Create batch metadata JSON files
                def metadata0 = file("batch_metadata_0.json")
                metadata0.text = '''{"sample_id": "sample1", "batch_id": 0, "kraken2_output": "test_batch0.kraken2.output.txt"}'''

                def metadata1 = file("batch_metadata_1.json")
                metadata1.text = '''{"sample_id": "sample1", "batch_id": 1, "kraken2_output": "test_batch1.kraken2.output.txt"}'''

                input[1] = [ metadata0, metadata1 ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("merge 3 batches - chronological order - stub") {
        options "-stub"

        when {
            process {
                """
                // Create 3 mock batch outputs (intentionally out of order)
                def batch2_output = file("test_batch2.kraken2.output.txt")
                batch2_output.text = "C\\tread5\\t11111\\t100\\t11111:1\\n"

                def batch0_output = file("test_batch0.kraken2.output.txt")
                batch0_output.text = "C\\tread1\\t12345\\t100\\t12345:1\\n"

                def batch1_output = file("test_batch1.kraken2.output.txt")
                batch1_output.text = "C\\tread3\\t67890\\t100\\t67890:1\\n"

                input[0] = [
                    [ id:'sample2' ],
                    [ batch2_output, batch0_output, batch1_output ]  // Intentionally unordered
                ]

                // Metadata files also unordered
                def metadata2 = file("batch_metadata_2.json")
                metadata2.text = '''{"sample_id": "sample2", "batch_id": 2, "kraken2_output": "test_batch2.kraken2.output.txt"}'''

                def metadata0 = file("batch_metadata_0.json")
                metadata0.text = '''{"sample_id": "sample2", "batch_id": 0, "kraken2_output": "test_batch0.kraken2.output.txt"}'''

                def metadata1 = file("batch_metadata_1.json")
                metadata1.text = '''{"sample_id": "sample2", "batch_id": 1, "kraken2_output": "test_batch1.kraken2.output.txt"}'''

                input[1] = [ metadata2, metadata0, metadata1 ]  // Intentionally unordered
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.cumulative_output },
                { assert process.out.stats },
                { assert snapshot(
                    process.out.cumulative_output,
                    process.out.stats,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("merge single batch - edge case - stub") {
        options "-stub"

        when {
            process {
                """
                def batch0_output = file("test_batch0.kraken2.output.txt")
                batch0_output.text = "C\\tread1\\t12345\\t100\\t12345:1\\nU\\tread2\\t0\\t100\\t0:100\\n"

                input[0] = [
                    [ id:'sample3' ],
                    [ batch0_output ]
                ]

                def metadata0 = file("batch_metadata_0.json")
                metadata0.text = '''{"sample_id": "sample3", "batch_id": 0, "kraken2_output": "test_batch0.kraken2.output.txt"}'''

                input[1] = [ metadata0 ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.cumulative_output },
                { assert process.out.stats }
            )
        }
    }

    test("merge 5 batches - large scale - stub") {
        options "-stub"

        when {
            process {
                """
                // Create 5 batch outputs
                def outputs = []
                def metadatas = []

                (0..4).each { idx ->
                    def batch_output = file("test_batch\${idx}.kraken2.output.txt")
                    batch_output.text = "C\\tread\${idx}\\t12345\\t100\\t12345:1\\n"
                    outputs << batch_output

                    def metadata = file("batch_metadata_\${idx}.json")
                    metadata.text = '''{"sample_id": "sample4", "batch_id": ''' + idx + ''', "kraken2_output": "test_batch''' + idx + '''.kraken2.output.txt"}'''
                    metadatas << metadata
                }

                input[0] = [
                    [ id:'sample4' ],
                    outputs
                ]
                input[1] = metadatas
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.cumulative_output },
                { assert process.out.stats }
            )
        }
    }

    test("metadata preservation - stub") {
        options "-stub"

        when {
            process {
                """
                def batch0_output = file("special_sample_batch0.kraken2.output.txt")
                batch0_output.text = "C\\tread1\\t12345\\t100\\t12345:1\\n"

                input[0] = [
                    [ id:'special_sample', single_end:true, run_id:'RUN001' ],
                    [ batch0_output ]
                ]

                def metadata0 = file("batch_metadata_0.json")
                metadata0.text = '''{"sample_id": "special_sample", "batch_id": 0, "kraken2_output": "special_sample_batch0.kraken2.output.txt"}'''

                input[1] = [ metadata0 ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.cumulative_output.get(0).get(0).id == 'special_sample' },
                { assert process.out.cumulative_output.get(0).get(0).single_end == true },
                { assert process.out.cumulative_output.get(0).get(0).run_id == 'RUN001' }
            )
        }
    }
}
