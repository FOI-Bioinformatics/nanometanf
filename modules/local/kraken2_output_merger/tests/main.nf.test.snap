{
    "merge 2 batches - single sample - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "sample1"
                        },
                        "sample1.cumulative.kraken2.output.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "sample1"
                        },
                        "merge_stats.json:md5,5d924393e7a07202e2e69cd2ad3cb6f9"
                    ]
                ],
                "2": [
                    "versions.yml:md5,9aa44ea7f9e936b45d354a13672c9b4f"
                ],
                "cumulative_output": [
                    [
                        {
                            "id": "sample1"
                        },
                        "sample1.cumulative.kraken2.output.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "sample1"
                        },
                        "merge_stats.json:md5,5d924393e7a07202e2e69cd2ad3cb6f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9aa44ea7f9e936b45d354a13672c9b4f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T16:56:45.873846"
    },
    "merge 3 batches - chronological order - stub": {
        "content": [
            [
                [
                    {
                        "id": "sample2"
                    },
                    "sample2.cumulative.kraken2.output.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "sample2"
                    },
                    "merge_stats.json:md5,5130d721e19ee37d0d3aaa47f9b9deee"
                ]
            ],
            [
                "versions.yml:md5,9aa44ea7f9e936b45d354a13672c9b4f"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T18:50:16.374354"
    }
}